baserun/instrumentation/openai.py

```diff
@@ -0,0 +1,187 @@
+                span.set_attribute(SpanAttributes.LLM_CHAT_STOP_SEQUENCES, [stop])
+            else:
+                span.set_attribute(SpanAttributes.LLM_CHAT_STOP_SEQUENCES, stop)
+
+        if "logit_bias" in kwargs:
+            span.set_attribute(
+                SpanAttributes.LLM_LOGIT_BIAS, json.dumps(kwargs.get("logit_bias"))
+            )
+
+        if "logprobs" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_LOGPROBS, kwargs.get("logprobs"))
+
+        if "echo" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_ECHO, kwargs.get("echo"))
+
+        if "suffix" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_SUFFIX, kwargs.get("suffix"))
+
+        if "best_of" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_BEST_OF, kwargs.get("best_of"))
+
+        if "user" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_USER, kwargs.get("user"))
+
+        messages = kwargs.get("messages", [])
+        for i, message in enumerate(messages):
+            prefix = f"{SpanAttributes.LLM_PROMPTS}.{i}"
+            span.set_attribute(f"{prefix}.role", message.get("role"))
+
+            if content := message.get("content"):
+                span.set_attribute(f"{prefix}.content", content)
+
+        if (prompt := kwargs.get("prompt")) and not messages:
+            span.set_attribute(f"{SpanAttributes.LLM_PROMPTS}.0.content", prompt)
+
+    @staticmethod
+    def set_response_attributes(span: Span, response: OpenAIObject):
+        choices = response.get("choices", [])
+        for i, choice in enumerate(choices):
+            prefix = f"{SpanAttributes.LLM_COMPLETIONS}.{i}"
+            span.set_attribute(f"{prefix}.finish_reason", choice.get("finish_reason"))
+
+            message = choice.get("message")
+            text = choice.get("text")
+            if message:
+                span.set_attribute(f"{prefix}.role", message.get("role"))
+
+                if content := message.get("content"):
+                    span.set_attribute(f"{prefix}.content", content)
+
+                if "function_call" in message:
+                    span.set_attribute(
+                        f"{prefix}.function_call",
+                        json.dumps(message.get("function_call").to_dict()),
+                    )
+
+            elif text:
+                span.set_attribute(f"{prefix}.content", text)
+
+        usage = response.get("usage")
+        if usage:
+            span.set_attribute(
+                SpanAttributes.LLM_USAGE_TOTAL_TOKENS, usage.get("total_tokens")
+            )
+            span.set_attribute(
+                SpanAttributes.LLM_USAGE_COMPLETION_TOKENS,
+                usage.get("completion_tokens"),
+            )
+            span.set_attribute(
+                SpanAttributes.LLM_USAGE_PROMPT_TOKENS,
+                usage.get("prompt_tokens"),
+            )
+
+    @staticmethod
+    def generator_wrapper(original_generator: GeneratorType, span: Span):
```

Comment:
I believe this needs to handle function calls as well

Reply:
I don't think streaming responses can have function calls https://platform.openai.com/docs/api-reference/chat/streaming

Reply:
If you expand a bit I think you'll see function_call in the delta object

<img width="1211" alt="Screenshot 2023-09-28 at 11 21 17 PM" src="https://github.com/baserun-ai/baserun-py/assets/125328015/a28c05fd-b3cc-4852-93ea-07796f5b459a">